chore: added route listener on non-trackable URLs

ankitsrivas14 opened the pull request:
Added route listener using NavigationSaga for non-trackable urls to track and send usage pulse.
>
>
#### PR fixes following issue(s)
Fixes #24892 
>
>
#### Media
> A video or a GIF is preferred. when using Loom, don’t embed because it looks like it’s a GIF. instead, just link to the video
>
>
#### Type of change
> Please delete options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update
>
>
>
## Testing
>
#### How Has This Been Tested?
> Please describe the tests that you ran to verify your changes. Also list any relevant details for your test configuration.
> Delete anything that is not relevant
- [x] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Test Plan
> Add Testsmith test cases links that relate to this PR
>
>
#### Issues raised during DP testing
> Link issues raised during DP testing for better visiblity and tracking (copy link from comments dropped on this PR)
>
>
>
## Checklist:
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag


#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

sbalaji1192 commented on app/client/src/ce/sagas/NavigationSagas.ts:
```diff
@@ -65,6 +67,16 @@ function* clearErrors() {
+    UsagePulse.isTrackableUrl,
+    window.location.pathname,
+  );
+  if (!isTrackable) {
```
We should call track when isTrackable is true right?

sbalaji1192 replied:
And also no need to check UsagePulse.isTrackableUrl here. track function checks that before sending pulse

ankitsrivas14 replied:
@sbalaji1192 - If we do not use `UsagePulse.isTrackableUrl` here then moving from Trackable URL to Trackable URL will also send a pulse, which means each time user clicks on a widget, a pulse would fire.

sbalaji1192 reviewed (commented):
My suggestion would be to keep the usage pulse as independent of the application as possible for better maintenance. But if we cannot figure out the problem with history listen or window pop state event is not working for then we can go this route.

ankitsrivas14 commented:
/ok-to-test

ankitsrivas14 commented:
/build-deploy-preview recreate=true

ankitsrivas14 commented:
/ok-to-test

ankitsrivas14 commented:
/ok-to-test

ankitsrivas14 commented:
/ok-to-test

ankitsrivas14 commented:
@sbalaji1192 - QA verified on https://github.com/appsmithorg/appsmith-ee/pull/1966#issuecomment-1664961205
Please check.

ankitsrivas14 commented:
@RakshaKShetty - Back Button and Usage Pulse QA for CE was also done, right? Please confirm.

RakshaKShetty commented:
@ankitsrivas14 yes its done from my side

ankitsrivas14 commented:
After discussing with @sbalaji1192 and @dipyamanbiswas07 regarding this ticket. We have decided to merge this PR as it will fix https://github.com/appsmithorg/appsmith/issues/24892 and the Usage pulse would remain as it is (QA verified).
We have created https://github.com/appsmithorg/appsmith/issues/25617 ticket to look into `history.listen` as per the bandwidth.

sbalaji1192 commented:
@ankitsrivas14 Can you please remove the dead code before merging the pr?

ankitsrivas14 commented:
I've removed the dead code and updated the logic to send pulse only when
1. App mounted on a trackable URL
2. Route changes from non-trackable URL to trackable URL

Please check @dipyamanbiswas07 and @sbalaji1192

ankitsrivas14 commented:
/ok-to-test

ankitsrivas14 commented:
/build-deploy-preview recreate=true

RakshaKShetty commented:
@ankitsrivas14 this looks good
